Sources/Core/Renderer/GUI/GUIElement/GUITextInput.swift

```diff
@@ -12,21 +13,43 @@ struct GUITextInput: GUIElement {
     ).meshes(context: context)
 
     // Message
-    var message = try content.meshes(context: context)
-    message.translate(amount: [2, 2])
+    var messageBeforeCursor = String(content.prefix(upTo: cursorIndex))
+    var messageAfterCursor = String(content.suffix(from: cursorIndex))
+    var messageMeshBeforeCursor = try messageBeforeCursor.meshes(context: context)
+    var messageMeshAfterCursor = try messageAfterCursor.meshes(context: context)
 
-    // Cursor
+    if messageMeshBeforeCursor.size().x != 0 {
+      messageMeshBeforeCursor.translate(amount: [2, 2])
+      messageMeshAfterCursor.translate(amount: [3 + messageMeshBeforeCursor.size().x, 2])
+    } else {
+      messageMeshBeforeCursor.translate(amount: [0, 2])
+      messageMeshAfterCursor.translate(amount: [2, 2])
+    }
     var cursor: [GUIElementMesh] = []
+    // When at the end of the message, use an underscore cursor. Otherwise, use a vertical bar cursor
     if Int(CFAbsoluteTimeGetCurrent() * 10/3) % 2 == 1 {
-      cursor = try "_".meshes(context: context)
-      let messageWidth = message.size().x
-      var xOffset = 2
-      if messageWidth != 0 {
-        xOffset = messageWidth + 4
+      if messageMeshAfterCursor.size().x == 0 {
+        cursor = try "_".meshes(context: context)
+        let messageWidth = messageMeshBeforeCursor.size().x + messageMeshAfterCursor.size().x
+        var xOffset = 2
+        if messageWidth != 0 {
+          xOffset = messageWidth + 4
+        }
```

Comment:
In Delta Client I generally prefer the following when conditionally initialising a variable from two possible values. It's a bit more code but I personally find it a bit more clear (and it also makes it clear that the variable isn't changed somewhere later in the code because you can use `let`). In newer Swift versions you can make the code nicer with if-expressions, but I haven't updated the project to the latest Swift version yet because of supporting older platforms.

```swift
let xOffset: Int
if messageWidth != 0 {
  xOffset = messageWidth + 4
} else {
  xOffset = 2
}
```

The same suggestion applies to lines 46 to 49.